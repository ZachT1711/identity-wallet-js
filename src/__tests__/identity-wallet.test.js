const IdentityWallet = require('../identity-wallet')
const HDNode = require('@ethersproject/hdnode')
const { verifyJWT } = require('did-jwt')
const { Resolver } = require('did-resolver')

const wallet1Conf = {
  seed: HDNode.mnemonicToSeed('clay rubber drama brush salute cream nerve wear stuff sentence trade conduct')
}
const wallet2Conf = {
  authSecret: '24a0bc3a2a1d1404c0ab24bef9bb0618938ee892fbf62f63f82f015eddf1729e'
}
const secondaryAuthSecret = '4567898765434567c0ab24bef9bb0618938ee892fbf62f63f82f015eddf1729e'
const badAuthData = [{
  nonce: 'Lxcd05Yk4aC8LCLbFjowzD3W6Uqx+v+n',
  ciphertext: 'elxT3d5Cxx4N9kIzRnJx0U1iKB1wLQu2u4pebshF3xXUEhw72rbCCfTsnNEKY3185MhRok0/t23Iyel5r6HJx/YOfj1XaKb4t9Ci8y21Bs38rQ=='
}]
const getConsentMock = jest.fn(() => false)


const didResolver = new Resolver({
  '3': async (_, { id }) => {
    let key
    if (id === 'bafyreia6evyez2xdlewmbh7hfz3dz3besmlhnlrnkiounscnnvboym7q2u' || id === 'first') {
      key = '027ab5238257532f486cbeeac59a5721bbfec2f13c3d26516ca9d4c5f0ec1aa229'
    } else {
      // key for 'space1'
      key = '0283441873077702f08a9e84d0ff869b5d08cb37361d77c7e5c57777e953670a0d'
    }
    return {
      '@context': 'https://w3id.org/did/v1',
      'id': 'did:3:' + id,
      'publicKey': [{
        'id': 'did:3:' + id + '#owner',
        'type': 'Secp256k1VerificationKey2018',
        'owner': 'did:3:' + id,
        'publicKeyHex': key
      }],
      'authentication': [{
        'type': 'Secp256k1SignatureAuthentication2018',
        'publicKey': 'did:3:' + id + '#owner'
      }]
    }
  }
})

const verifyJWTOpts = { resolver: didResolver }

//registerMethod('3', async (_, { id }) => {
  //let key
  //if (id === 'bafyreia6evyez2xdlewmbh7hfz3dz3besmlhnlrnkiounscnnvboym7q2u' || id === 'first') {
    //key = '027ab5238257532f486cbeeac59a5721bbfec2f13c3d26516ca9d4c5f0ec1aa229'
  //} else {
    //// key for 'space1'
    //key = '0283441873077702f08a9e84d0ff869b5d08cb37361d77c7e5c57777e953670a0d'
  //}
  //return {
    //'@context': 'https://w3id.org/did/v1',
    //'id': 'did:3:' + id,
    //'publicKey': [{
      //'id': 'did:3:' + id + '#owner',
      //'type': 'Secp256k1VerificationKey2018',
      //'owner': 'did:3:' + id,
      //'publicKeyHex': key
    //}],
    //'authentication': [{
      //'type': 'Secp256k1SignatureAuthentication2018',
      //'publicKey': 'did:3:' + id + '#owner'
    //}]
  //}
//})


describe('IdentityWallet', () => {

  let idWallet1, idWallet2

  beforeAll(() => {
    idWallet1 = new IdentityWallet(getConsentMock, wallet1Conf)
    idWallet2 = new IdentityWallet(getConsentMock, wallet2Conf)
  })

  it('should be correctly constructed', async () => {
    expect(idWallet1._seed).toBeDefined()
    expect(idWallet2._authSecret).toBeDefined()
  })

  it('getLink correctly', async () => {
    const linkProofPromise = new Promise((resolve, reject) => {
      idWallet1.events.on('new-link-proof', resolve)
    })
    expect(await idWallet1.getLink()).toMatchSnapshot()
    await linkProofPromise
    expect(await idWallet2.getLink()).toMatchSnapshot()
  })

  describe('consent functionality', () => {
    const origin = 'https://my.origin'
    const type = 'authenticate'

    beforeEach(() => {
      getConsentMock.mockClear()
    })

    it('should throw if getConsent param not passed to constructor', async () => {
      expect(() => new IdentityWallet(wallet1Conf)).toThrow('getConsent parameter has to be a function')
    })

    it('returns false if no consent given', async () => {
      expect(await idWallet1.getConsent([], origin)).toBeFalsy()
      expect(getConsentMock).toHaveBeenCalledTimes(1)
      expect(getConsentMock).toHaveBeenCalledWith({ type, spaces: [], origin })
    })

    it('works without spaces', async () => {
      getConsentMock.mockImplementationOnce(() => true)
      expect(await idWallet1.getConsent([], origin)).toBeTruthy()
      expect(getConsentMock).toHaveBeenCalledTimes(1)
      expect(getConsentMock).toHaveBeenCalledWith({ type, spaces: [], origin })
    })

    it('should not call consent fn if consent already given', async () => {
      expect(await idWallet1.getConsent([], origin)).toBeTruthy()
      expect(getConsentMock).toHaveBeenCalledTimes(0)
    })

    it('should not have consent for different origin', async () => {
      expect(await idWallet1.hasConsent([], 'https://my.other.origin')).toBeFalsy()
    })

    it('works with spaces', async () => {
      const spaces = ['s1', 's2', 's3']
      getConsentMock.mockImplementationOnce(() => true)
      expect(await idWallet1.getConsent(spaces, origin)).toBeTruthy()
      expect(getConsentMock).toHaveBeenCalledTimes(1)
      expect(getConsentMock).toHaveBeenCalledWith({ type, spaces, origin })
    })

    it('works with spaces, already have consent', async () => {
      const spaces = ['s1', 's2']
      expect(await idWallet1.getConsent(spaces, origin)).toBeTruthy()
      expect(getConsentMock).toHaveBeenCalledTimes(0)
    })
  })

  it('addAuthMethod should throw before authenticate', async () => {
    await expect(idWallet2.addAuthMethod()).rejects.toMatchSnapshot()
  })

  describe('authenticate', () => {

    let authPubKeys
    let authData = []

    beforeEach(() => {
      getConsentMock.mockImplementation(() => true)
      idWallet2 = new IdentityWallet(getConsentMock, wallet2Conf)
    })

    it('works correctly w/ seed only wallet', async () => {
      expect(await idWallet1.authenticate()).toMatchSnapshot()
      expect(await idWallet1.authenticate(['space1', 'space2'])).toMatchSnapshot()
      expect(await idWallet1.authenticate(['space3', 'space4'])).toMatchSnapshot()
      expect(await idWallet1.authenticate(['space2', 'space3'])).toMatchSnapshot()
      expect(await idWallet1.authenticate([], { mgmtPub: true })).toMatchSnapshot()
    })

    it('should generate seed if no auth data passed', async () => {
      const authDataPromise = new Promise((resolve, reject) => {
        idWallet2.events.on('new-auth-method', resolve)
      })
      const linkProofPromise = new Promise((resolve, reject) => {
        idWallet2.events.on('new-link-proof', resolve)
      })
      expect(idWallet2._keyring).toBeUndefined()
      authPubKeys = await idWallet2.authenticate()
      authData.push(await authDataPromise)
      await linkProofPromise
    })

    it('should auth if auth-data is passed', async () => {
      expect(idWallet2._keyring).toBeUndefined()
      expect(await idWallet2.authenticate([], { authData })).toEqual(authPubKeys)
    })

    it('should throw if no valid authSecrets', async () => {
      await expect(idWallet2.authenticate([], { authData: badAuthData })).rejects.toMatchSnapshot()
    })

    it('addAuthMethod works correctly', async () => {
      const authDataPromise = new Promise((resolve, reject) => {
        idWallet2.events.on('new-auth-method', resolve)
      })
      expect(await idWallet2.authenticate([], { authData })).toEqual(authPubKeys)
      const linkProofPromise = new Promise((resolve, reject) => {
        idWallet2.events.on('new-link-proof', resolve)
      })
      await idWallet2.addAuthMethod(secondaryAuthSecret)
      authData.push(await authDataPromise)
      const linkAddress = (await linkProofPromise).address

      const idWallet3 = new IdentityWallet(getConsentMock, {
        authSecret: secondaryAuthSecret
      })
      expect((await idWallet3.getLink()).toLowerCase()).toEqual(linkAddress)
      expect(await idWallet3.authenticate([], { authData })).toEqual(authPubKeys)
    })
  })

  it('signClaim creates JWTs correctly', async () => {
    const payload = {
      some: 'data'
    }
    const jwt0 = await idWallet1.signClaim(payload)
    const jwt1 = await idWallet1.signClaim(payload, { DID: 'did:3:first' })
    const jwt2 = await idWallet1.signClaim(payload, { space: 'space1' })

    expect(await verifyJWT(jwt0, verifyJWTOpts)).toBeDefined()
    expect(await verifyJWT(jwt1, verifyJWTOpts)).toBeDefined()
    expect(await verifyJWT(jwt2, verifyJWTOpts)).toBeDefined()
  })

  it('encrypt/decrypt works correctly', async () => {
    // encrypt and decrypt should work
    const msg1 = 'secret message'
    const encObj1 = await idWallet1.encrypt(msg1)
    expect(await idWallet1.decrypt(encObj1)).toEqual(msg1)

    // decrypt with wrong key should fail
    await expect(idWallet1.decrypt(encObj1, 'space1')).rejects.toMatchSnapshot()

    // encrypt and decrypt should work with space
    const msg2 = 'secret message two'
    const encObj2 = await idWallet1.encrypt(msg2, 'space1')
    expect(await idWallet1.decrypt(encObj2, 'space1')).toEqual(msg2)
  })

  it('asymmetrically encrypt/decrypt works correctly', async () => {
    // encrypt and decrypt should work
    const msg1 = 'secret message'
    const { asymEncryptionKey } = idWallet1._keyring.getPublicKeys({ space: 'space1' })
    const encObj1 = await idWallet1.encrypt(msg1, null, { to: asymEncryptionKey })
    expect(await idWallet1.decrypt(encObj1, 'space1')).toEqual(msg1)

    // decrypt with wrong key should fail
    await expect(idWallet1.decrypt(encObj1, 'space2')).rejects.toMatchSnapshot()
  })
})
