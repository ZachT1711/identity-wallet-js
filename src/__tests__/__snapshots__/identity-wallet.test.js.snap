// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdentityWallet addAuthMethod should throw before authenticate 1`] = `[Error: This method can only be called after authenticate has been called]`;

exports[`IdentityWallet asymmetrically encrypt/decrypt works correctly 1`] = `[Error: IdentityWallet: Could not decrypt message]`;

exports[`IdentityWallet authenticate should throw if no valid authSecrets 1`] = `[Error: No valid auth-secret for this identity]`;

exports[`IdentityWallet authenticate works correctly w/ seed only wallet 1`] = `
Object {
  "main": Object {
    "asymEncryptionKey": "jRKRy3oPhXpUBrFJefO2CJnWw/IoalOPuepAwvTqrEk=",
    "managementKey": "0xa7b76379659b6c19E4DFf46e0DfEa48AFEBA8fe3",
    "signingKey": "027ab5238257532f486cbeeac59a5721bbfec2f13c3d26516ca9d4c5f0ec1aa229",
  },
  "spaces": Object {},
}
`;

exports[`IdentityWallet authenticate works correctly w/ seed only wallet 2`] = `
Object {
  "main": Object {
    "asymEncryptionKey": "jRKRy3oPhXpUBrFJefO2CJnWw/IoalOPuepAwvTqrEk=",
    "managementKey": "0xa7b76379659b6c19E4DFf46e0DfEa48AFEBA8fe3",
    "signingKey": "027ab5238257532f486cbeeac59a5721bbfec2f13c3d26516ca9d4c5f0ec1aa229",
  },
  "spaces": Object {
    "space1": Object {
      "asymEncryptionKey": "aA5XCwjYEv6RhS11GzYb902nsdZns66orubcWGxnm2c=",
      "managementKey": null,
      "signingKey": "0483441873077702f08a9e84d0ff869b5d08cb37361d77c7e5c57777e953670a0debc728f18dc49203596a0b2392d8675bc28d8da62262350123c2c8ca4e709568",
    },
    "space2": Object {
      "asymEncryptionKey": "G/pdGl6OtWv1Vz5LplLZIJxOhH2AmXnF+2BYF5e1TUc=",
      "managementKey": null,
      "signingKey": "04836712d05ad740972438b48b8c88571fa387c856f1c73a9abed17586de49c5dd9c681bc296087f2f5831813b93b15311c2fe483c284ef02d6115fe4a20b2256e",
    },
  },
}
`;

exports[`IdentityWallet authenticate works correctly w/ seed only wallet 3`] = `
Object {
  "main": Object {
    "asymEncryptionKey": "jRKRy3oPhXpUBrFJefO2CJnWw/IoalOPuepAwvTqrEk=",
    "managementKey": "0xa7b76379659b6c19E4DFf46e0DfEa48AFEBA8fe3",
    "signingKey": "027ab5238257532f486cbeeac59a5721bbfec2f13c3d26516ca9d4c5f0ec1aa229",
  },
  "spaces": Object {
    "space3": Object {
      "asymEncryptionKey": "F4OJwO3rpESb+/D8Q8IKGzaNwQJ+TwHt+QOzEtSvkSk=",
      "managementKey": null,
      "signingKey": "046006731db949141b42b64bf1121be029bb5abe9e8267413db1ee9d6c6010507f51607253f839ac3a2ce2c064fca8049b3c4204f7e502ba6b56a961b5a309de4f",
    },
    "space4": Object {
      "asymEncryptionKey": "3QRW9cArRuXgIb0zI/r3/oaoP52TkIWc12Vr52XwAAE=",
      "managementKey": null,
      "signingKey": "047ed187e7ce1f88e6fdd058882cf6f48301efc4992697c623e7fb2837d8d8d8c0cadc19cf5c628c9034943a0ff48b591a1dc0ae83f40878b8b41da4d9bbd98bc6",
    },
  },
}
`;

exports[`IdentityWallet authenticate works correctly w/ seed only wallet 4`] = `
Object {
  "main": Object {
    "asymEncryptionKey": "jRKRy3oPhXpUBrFJefO2CJnWw/IoalOPuepAwvTqrEk=",
    "managementKey": "0xa7b76379659b6c19E4DFf46e0DfEa48AFEBA8fe3",
    "signingKey": "027ab5238257532f486cbeeac59a5721bbfec2f13c3d26516ca9d4c5f0ec1aa229",
  },
  "spaces": Object {
    "space2": Object {
      "asymEncryptionKey": "G/pdGl6OtWv1Vz5LplLZIJxOhH2AmXnF+2BYF5e1TUc=",
      "managementKey": null,
      "signingKey": "04836712d05ad740972438b48b8c88571fa387c856f1c73a9abed17586de49c5dd9c681bc296087f2f5831813b93b15311c2fe483c284ef02d6115fe4a20b2256e",
    },
    "space3": Object {
      "asymEncryptionKey": "F4OJwO3rpESb+/D8Q8IKGzaNwQJ+TwHt+QOzEtSvkSk=",
      "managementKey": null,
      "signingKey": "046006731db949141b42b64bf1121be029bb5abe9e8267413db1ee9d6c6010507f51607253f839ac3a2ce2c064fca8049b3c4204f7e502ba6b56a961b5a309de4f",
    },
  },
}
`;

exports[`IdentityWallet authenticate works correctly w/ seed only wallet 5`] = `
Object {
  "main": Object {
    "asymEncryptionKey": "jRKRy3oPhXpUBrFJefO2CJnWw/IoalOPuepAwvTqrEk=",
    "managementKey": "0350966498e802a007513d518dc12a0a0ab1e48addcec1beaa989c6ff5f5ed226c",
    "signingKey": "027ab5238257532f486cbeeac59a5721bbfec2f13c3d26516ca9d4c5f0ec1aa229",
  },
  "spaces": Object {},
}
`;

exports[`IdentityWallet encrypt/decrypt works correctly 1`] = `[Error: IdentityWallet: Could not decrypt message]`;

exports[`IdentityWallet getLink correctly 1`] = `"0xa7b76379659b6c19E4DFf46e0DfEa48AFEBA8fe3"`;

exports[`IdentityWallet getLink correctly 2`] = `"0xE72E6c37c15986C86Aa94007451750176E9a999F"`;
